src/software/embedded/robot_diagnostics_cli/embedded_communication.py

```diff
@@ -0,0 +1,162 @@
+
+    BROADCAST_HZ = 1.0
+    LOCALHOST_IP = "127.0.0.1"
+
+    def __init__(self):
+        """Establishes multicast connections and e-stop information"""
+        self.primitive_sender: tbots_cpp.PrimitiveSender | None = None
+        self.receive_robot_status: tbots_cpp.RobotStatusProtolistener | None = None
+        self.embedded_data = EmbeddedData()
+        self.channel_id = int(self.embedded_data.get_channel_id())
+        self.robot_id = int(self.embedded_data.get_robot_id())
+
+        self.sequence_number = 0
+        self.command_duration_seconds = 2.0
+        self.send_primitive_interval_s = 0.01
+
+        # total primitives for this robots
+        self.robot_primitives = {}
+        self.running = True
+        # Set the primitive of the local robot
+        self.robot_primitives[self.robot_id] = Primitive(stop=StopPrimitive())
+
+        # Localhost IP Broadcaster
+        self.fullsystem_ip_broadcaster = tbots_cpp.FullsystemIpBroadcastProtoUdpSender(
+            getRobotMulticastChannel(int(self.embedded_data.get_channel_id())),
+            FULL_SYSTEM_TO_ROBOT_IP_NOTIFICATION_PORT,
+            self.embedded_data.get_network_interface(),
+            True,
+        )
+
+        # initialising the estop
+        self.estop_reader = None
+        self.should_send_stop = False
+        self.estop_mode, self.estop_path = get_estop_config(
+            keyboard_estop=False, disable_communication=False
+        )
+
+        # Always in PHYSICAL_ESTOP mode within CLI
+        try:
+            self.estop_reader = tbots_cpp.ThreadedEstopReader(self.estop_path, 115200)
+            self.__update_estop_state()
+        except Exception as e:
+            raise Exception(f"Invalid Estop found at location {self.estop_path} as {e}")
+
+    def __receive_robot_status(self, robot_status: Message) -> None:
+        """Updates the dynamic information with the new RobotStatus message.
+        :param robot_status: The incoming RobotStatus proto from the robot to read
+        """
+        self.embedded_data.epoch_timestamp_seconds = (
+            robot_status.time_sent.epoch_timestamp_seconds
+        )
+        self.embedded_data.battery_voltage = robot_status.power_status.battery_voltage
+        self.embedded_data.primitive_packet_loss_percentage = (
+            robot_status.network_status.primitive_packet_loss_percentage
+        )
+        self.embedded_data.primitive_executor_step_time_ms = (
+            robot_status.thunderloop_status.primitive_executor_step_time_ms
+        )
+
+    def __should_send_packet(self) -> bool:
+        """Returns True if a proto should be sent
+        :return: boolean
+        """
+        return (
+            self.estop_mode != EstopMode.DISABLE_ESTOP
+        ) and not self.should_send_stop
+
+    def send_primitive(self, primitive: Primitive) -> None:
+        """Forward PrimitiveSet protos from diagnostics to the robots."""
```

Comment:
We should avoid mentioning PrimitiveSet in docs since PrimitiveSet doesn't exist anymore